README.rst
=================================
e.bash: tiny expression evaluator
=================================

**e.bash** is a Bash loadable extension, just like ``printf`` or ``read``, it
runs fast. It's a fork of |ee|__, which was made for calculation with float
numbers in 2001.

.. |ee| replace:: *e*
__ History_

.. contents:: **Contents**
   :local:


Who should use e.bash?
======================

No one, and no one should use ``bc``, too, if you are doing mass calculations
with float numbers in shell scripting. If you are doing that, then you are
doing it wrong. Use another shell that supports float natively, or the better,
use another programming language for that kind of task.

How is it wrong? Have you seen the numbers in Benchmark_? How much processing
power have you wasted? Bash isn't meant for mass float calculation, in fact, it
can't. That's why many people use external ``bc``, and most of them ain't aware
of how costly to invoke external command.

Nevertheless, if you just want to say "Because I can," well, go ahead.

Or just want to get ``1 + 1 = ?``, then either programs are fine to use.


Installation
============

By default, it's installed to ``/usr/local`` and you will need to provide the
path to Bash header files using ``BASH_INC``:

.. code:: sh

  $ make [prefix=PREFIX] BASH_INC=/path/to/bash/include [install|install-strip]

You can change the installed location using ``prefix``. If everything compiles
successfully, you should have two binary files:

* ``e``: the command
* ``e.bash``: the loadable

Both are installed to ``prefix/bin``, although you can't directly execute
``e.bash``, which is meant to be loaded by Bash, see Loadable_.

If you need to strip, use ``install-strip`` target to install. To uninstall,
with same ``prefix``, use ``uninstall`` target.


Usage
=====

.. code:: sh

  $ e [OPTIONS] [--] [expression]

``-h``:
  display help message

``-V``:
  display version string

``-d``:
  display debugging messages

``-v varname`` (loadable only):
  store result to variable ``varname``

Loadable
--------

.. code:: sh

  $ enable -f /path/to/e.bash e
  $ e [expression]

The result will be stored (Variable Binding) to ``$VARNAME``, of course, you
can use with Command Substitution:

.. code:: sh

  $ echo $(e [expression])

Command
-------

.. code:: sh

  $ ./e [expression]


Expression
==========

Operators
---------

+-------------------+--------------------------+---------------+------------+
| operators         | explanation              | evaluation    | precedence |
+===================+==========================+===============+============+
| ``+`` ``-``       | add,subtract             | left to right | lower      |
+-------------------+--------------------------+---------------+------------+
| ``*`` ``/`` ``%`` | multiply, divide, modulo | left to right |            |
+-------------------+--------------------------+---------------+------------+
| ``^``             | exponentiate             | right to left |            |
+-------------------+--------------------------+---------------+------------+
| ``!``             | factorial                | obvious       | higher     |
+-------------------+--------------------------+---------------+------------+

Constants
---------

mathematical
............

+-------------+----------------------------+--------------+
| constants   | value                      | in C         |
+=============+============================+==============+
| ``e``       | ``2.7182818284590452354``  | ``M_E``      |
+-------------+------------------------+------------------+
| ``pi``      | ``3.14159265358979323846`` | ``M_PI``     |
+-------------+------------------------+------------------+

limitations
...........

+-------------+------------------------+--------------+
| constants   | value                  | in C         |
+=============+========================+==============+
| ``dblmin``  | *too long to display*  | ``DBL_MIN``  |
+-------------+------------------------+--------------+
| ``dblmax``  | *too long to display*  | ``DBL_MAX``  |
+-------------+------------------------+--------------+
| ``randmax`` | ``2147483647``         | ``RAND_MAX`` |
+-------------+------------------------+--------------+

Functions
---------

sign and absolute
.................

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``sign``      | sign of number                                            |
+---------------+-----------------------------------------------------------+
| ``abs``       | absolute value of integer number                          |
+---------------+-----------------------------------------------------------+
| ``fabs``      | absolute value of floating point number                   |
+---------------+-----------------------------------------------------------+

exponents and logarithms
........................

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``exp``       | base e exponential                                        |
+---------------+-----------------------------------------------------------+
| ``ln``        | natural logarithm (base e)                                |
+---------------+-----------------------------------------------------------+
| ``log``       | base 2 logarithm (log256 = 8)                             |
+---------------+-----------------------------------------------------------+
| ``sqrt``      | sqrt                                                      |
+---------------+-----------------------------------------------------------+

rounding
........

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``floor``     | largest integral value, not greater than argument         |
+---------------+-----------------------------------------------------------+
| ``ceil``      | smallest integral, not less than argument                 |
+---------------+-----------------------------------------------------------+
| ``round``     | round to nearest integer, away from zero                  |
+---------------+-----------------------------------------------------------+
| ``trunc``     | round to integer, toward zero                             |
+---------------+-----------------------------------------------------------+

trigonometric
.............

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``sin``       | sine                                                      |
+---------------+-----------------------------------------------------------+
| ``cos``       | cosine                                                    |
+---------------+-----------------------------------------------------------+
| ``tan``       | tangent                                                   |
+---------------+-----------------------------------------------------------+
| ``asin``      | arc sine                                                  |
+---------------+-----------------------------------------------------------+
| ``acos``      | arc cosine                                                |
+---------------+-----------------------------------------------------------+
| ``atan``      | arc tangent                                               |
+---------------+-----------------------------------------------------------+

hyperbolic
..........

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``sinh``      | hyperbolic sine                                           |
+---------------+-----------------------------------------------------------+
| ``cosh``      | hyperbolic cosine                                         |
+---------------+-----------------------------------------------------------+
| ``tanh``      | hyperbolic tangent                                        |
+---------------+-----------------------------------------------------------+
| ``asinh``     | inverse hyperbolic sine                                   |
+---------------+-----------------------------------------------------------+
| ``acosh``     | inverse hyperbolic cosine                                 |
+---------------+-----------------------------------------------------------+
| ``atanh``     | inverse hyperbolic tangent                                |
+---------------+-----------------------------------------------------------+

random
......

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``srand``     | set random seed, note that it returns value of 0          |
|               | due to the current design of e.bash                       |
+---------------+-----------------------------------------------------------+
| ``rand``      | integer random number,  in range 0 to ``RAND_MAX``,       |
|               | inclusively                                               |
+---------------+-----------------------------------------------------------+
| ``randf``     | shorthand for ``rand / randmax``                          |
+---------------+-----------------------------------------------------------+

miscellaneous
.............

+---------------+-----------------------------------------------------------+
| functions     | explanation                                               |
+===============+===========================================================+
| ``time``      | timestamp in seconds since Unix epoch                     |
+---------------+-----------------------------------------------------------+

Benchmark
=========

.. code:: sh

  $ make benchmark

Sample result:

+-----------------+--------+
| method          | runs   |
+=================+========+
| original ``e``  | 589    |
+-----------------+--------+
| ``e``           | 539    |
+-----------------+--------+
| loadable        | 985    |
+-----------------+--------+
| loadable ``-v`` | 15,763 |
+-----------------+--------+

Sample result for 1 + 1:

+-----------------------+--------+
| method                | runs   |
+=======================+========+
| Arithmetic Expansion  | 21,336 |
+-----------------------+--------+
| Arithmetic Evaluation | 22,467 |
+-----------------------+--------+
| let                   | 21,144 |
+-----------------------+--------+
| loadable ``-v``       | 17,398 |
+-----------------------+--------+

Examples
--------

+----------------+----------------+
| script         | time (seconds) |
+================+================+
| ``sine.sh``    | 0.028          |
+----------------+----------------+
| ``sine.bc.sh`` | 0.407          |
+----------------+----------------+


History
=======

The original |e|_ was written by Dimitromanolakis Apostolos in 2001, the
version 0.02718_ was released on 2011-07-11. From the original website e_:

.. |e| replace:: **e**
.. _e: http://web.archive.org/web/20090924080521/http://www.softnet.tuc.gr/%7Eapdim/projects/e/
.. _0.02718: https://bitbucket.org/livibetter/e.bash/commits/tag/v0.02718

  Some time ago while I [Dimitromanolakis Apostolos] was doing some homework
  for my university class, I needed a quick way to evaluate expressions, while
  I was typing at the command prompt. I found two solutions, using bc or
  gnuplot. bc has fixed precision which defaults to 0, so to evaluate an
  expression involving decimal results you need to issue a command like
  "scale=5" beforehand. On the other hand using gnuplot (and bc if it matters)
  involves loading the executable, evaluating your expression using the "print"
  command and quitting using the "quit" command. I needed something quicker..

  ...so, I coded *e*.

  e is a command line expression evaluator. It was designed to be as small as
  possible, and quick to use. Therefore the name "e" was chosen, so that while
  you are at the command prompt you can evaluate an expression with only 2
  keystrokes overhead. e manages to be under 8k in size on most compilers that
  I tried. My current record is 7000 bytes for v0.02718.

In August, 2014, e was forked and transformed into a Bash loadable extension by
Yu-Jie Lin (@livibetter) on Bitbucket.


Copyright
=========

This project is licensed under the GNU General Public License Version 2, see
COPYING_.

.. _COPYING: COPYING


NEWS
===========================
NEWS for 0.4.0 (2015-07-01)
===========================

Changes since 0.3.0 (2014-09-01)
================================

[both]

* add function sign


Changes since 0.2.0 (2014-08-29)
================================

[both]

* add constant dblmin
* add constant dblmax
* add function round
* add function srand for setting random seed, it returns value of 0
* add function time for Unix timestamp
* add function trunc
* add -d for debugging message


Changes since 0.1.0 (2014-08-21)
================================

[both]

* add constant randmax for maximum integer for ineteger random number
* add function rand for integer random number
* add function randf for float random number, shorthand for rand / randmax
* add -h for help message

[command]

* add -V for version string

[examples]

* add f.sh: plot Y = F(X)
* add f-examples.sh: a set of examples of f.sh


Changes since 0.02718 (2001-07-11)
==================================

* compile as c99 for command and gnu99 for loadable
* make e to be both command and Bash loadable
* make possible to install and uninstall, and with prefix

[loadable]

* add -v to store result to the variable
* add -V to display version string

[examples]

* add sine.sh and sine.bc.sh


Changes since 0.0.0 (2001-07-11)
================================

* first public release


check.sh
#!/bin/bash

E2718="$PWD/e-0.02718/e"
EPROG="$PWD/e"
EBASH="$PWD/e.bash"

SKIP=0
PASS=0
TOTAL=0

FMT () {
  echo "$@" | sed '2,$ s/^/       /'
}

TEST () {
  TEST_F _TEST_EQ "$@"
}

TEST_F() {
  ((TOTAL++))
  _TEST "$@" && ((PASS++))
  SKIP_E2718=
}

_TEST () {
  local F T E RE2718 REPROG REBASH ERRORS

  F="$1"
  shift
  E="$1"    # expected result
  shift
  T=("$@")  # test input

  RE2718="$("$E2718" "${T[@]}")"
  REPROG="$("$EPROG" "${T[@]}")"
  REBASH="$( (enable -f "$EBASH" e && e "${T[@]}") )"

  ERRORS=0
  if [[ $SKIP_E2718 ]]; then
    ((SKIP++))
  else
    "$F" "$E" "$RE2718" || ((ERRORS++))
  fi
  "$F" "$E" "$REPROG" || ((ERRORS++))
  "$F" "$E" "$REBASH" || ((ERRORS++))

  ((ERRORS == 0)) && return 0

  FMT "INPUT: ${T[@]}"
  FMT "EXPCT: $E"
  [[ $SKIP_E2718 ]]   ||
  "$F" "$E" "$RE2718" || FMT "E2718: $RE2718"
  "$F" "$E" "$REPROG" || FMT "EPROG: $REPROG"
  "$F" "$E" "$REBASH" || FMT "EBASH: $REBASH"
  echo

  return 1
}

_TEST_EQ () {
  [[ $1 == $2 ]]
}

_TEST_RAND_10 () {
  ((0 <= $2)) && (($2 < 10))
}

# check Unix timestamp within range of 1 second
# 1 second should be enough when seconds crossing between calls of time from e
# and test function
_TEST_TIME () {
  local t

  printf -v t '%(%s)T' -1
  ((t == $2)) || (($2 <= t - 1))
}

_TEST_VERSION () {
  local VERSION D

  VERSION="e $(<version.h egrep -o '[0-9]+\.[0-9]+\.[0-9]+')"
  D=$(head -1 <<< "$2")

  [[ $VERSION == $D ]]
}

#########
# basic #
#########

TEST 0 0
TEST 1 1
TEST -1 -- -1
TEST 3.14159265358979 pi
TEST 1 sin[pi/2]

###################
# space-separated #
###################

TEST 3 '1+2'
TEST 3 1 + 2
TEST 3 1+ 2
TEST 3 1 +2
TEST 6 1 +2 + 3

###########
# min/max #
###########

SKIP_E2718=1
TEST "\
0.000000000000000000000000000000000000000000000000000000000000000000000000000\
00000000000000000000000000000000000000000000000000000000000000000000000000000\
00000000000000000000000000000000000000000000000000000000000000000000000000000\
00000000000000000000000000000000000000000000000000000000000000000000000000000\
022250738585072"\
     dblmin
SKIP_E2718=1
TEST "\
17976931348623157081452742373170435679807056752584499659891747680315726078002\
85387605895586327668781715404589535143824642343213268894641827684675467035375\
16986049910576551282076245490090389328944075868508455133942304583236903222948\
16580855933212334827479782620414472316873817718091929988125040402618412485836\
8"\
     dblmax

#############
# functions #
#############

# round
#######

SKIP_E2718=1
TEST  1 round[ 0.5]
SKIP_E2718=1
TEST -1 round[-0.5]

# sign
######

SKIP_E2718=1
TEST  0 sign[ 0]
SKIP_E2718=1
TEST  1 sign[ 1]
SKIP_E2718=1
TEST -1 sign[-1]

# trunc
#######

SKIP_E2718=1
TEST  1    trunc[ 1.5]
SKIP_E2718=1
TEST -1 -- trunc[-1.5]
SKIP_E2718=1
TEST  1    trunc[ 1  ]
SKIP_E2718=1
TEST -1 -- trunc[-1  ]

# time
######

SKIP_E2718=1
TEST_F _TEST_TIME ? time

################
# syntax error #
################

TEST $'123456789012345pi\n               ^--- syntax error' 123456789012345pi
SKIP_E2718=1
TEST $'1234567890123456pi\nsyntax error ---^' 1234567890123456pi
TEST $'12345678901234567pi\n syntax error ---^' 12345678901234567pi

##########
# random #
##########

SKIP_E2718=1
TEST 2147483647 randmax
SKIP_E2718=1
TEST 0 floor[randf*1]
SKIP_E2718=1
TEST_F _TEST_RAND_10 '?' floor[randf*10]

# srand
#######

SKIP_E2718=1
TEST 1205554746 srand[3] +rand
SKIP_E2718=1
TEST 1205554746 srand[pi]+rand
SKIP_E2718=1
TEST 2411109492 srand[pi]+rand+srand[pi]+rand
SKIP_E2718=1
TEST 1688702731 srand[pi]+rand+rand

##################
# version string #
##################

SKIP_E2718=1
TEST_F _TEST_VERSION '?' -V

echo

echo "PASS: $PASS of $TOTAL (SKIP = $SKIP)"
exit $((PASS < TOTAL))


benchmark.sh
#!/bin/bash
# Copyright (C) 2014  Yu-Jie Lin
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

: ${DURATION:=5}
: ${TEST:=sin[pi/2]}
: ${EXPECT:=1}
: ${VARNAME=RESULT}

trap 'kill $tmpcount_pid ; rm "$tmpcount" ; exit 130' INT

test_command () {

  local RESULT
  local VIA="$1"
  local TD="$2"
  local SETUP="$3"

  echo -ne "\e[1;37mBenchmarking\e[0m "
  [[ ! -z "${SETUP[@]}" ]] && echo -ne "\e[1;32m${SETUP[@]/$PWD/\$PWD}\e[0m; "
  echo -e "\e[1;33m${TD[@]}\e[0m..."

  echo -n "Please wait for $DURATION seconds..."
  tmpcount="$(mktemp)"
  (
    trap exit TERM
    eval "$SETUP"
    while RESULT=; do
      "$TD"
      [[ $RESULT == $EXPECT ]] && echo >> "$tmpcount"
    done
  ) &>/dev/null &
  tmpcount_pid=$!
  sleep $DURATION
  if ! kill $tmpcount_pid; then
    echo -e "\e[1;31msubprocess ($tmpcount_pid) already gone," \
            "something went wrong\e[0m" >&2
    exit 1
  fi

  RUNS=$(bc <<< "$(wc -l < "$tmpcount") / $DURATION")
  printf "\e[3K\e[0G\e[1;34m%'7.0f\e[0m runs per second via %s\n" "$RUNS" "$VIA"
  rm "$tmpcount"

  if ((RUNS == 0)); then
    echo -e "\e[1;31m0 runs, something went wrong\e[0m" >&2
    exit 1
  fi
}

#########
# basic #
#########

test_e2718 () {
  RESULT="$(e-0.02718/e $TEST)"
}

test_command 'original e program' test_e2718

test_e_command () {
  RESULT="$(./e $TEST)"
}

test_command 'e program' test_e_command

test_e_loadable () {
  RESULT="$(e $TEST)"
}

test_command 'Bash loadable with Command Substitution' \
             test_e_loadable \
             'enable -f $PWD/e.bash e'

test_e_loadable_v () {
  e -v RESULT $TEST
}

test_command 'Bash loadable with Variable Binding' \
             test_e_loadable_v \
             'enable -f $PWD/e.bash e'

#########
# 1 + 1 #
#########

EXPECT=2

test_bash_1plus1_ap () {
  RESULT="$((1 + 1))"
}

test_command 'Bash 1 + 1 Arithmetic Expansion' test_bash_1plus1_ap

test_bash_1plus1_ae () {
  ((RESULT=1+1))
}

test_command 'Bash 1 + 1 Arithmetic Evaluation' test_bash_1plus1_ae

test_bash_1plus1_let () {
  let RESULT=1+1
}

test_command 'Bash 1 + 1 let' test_bash_1plus1_let

test_e_loadable_v_1plus1 () {
  e -v RESULT 1 + 1
}

test_command 'Bash loadable with Variable Binding, 1 + 1' \
             test_e_loadable_v_1plus1 \
             'enable -f $PWD/e.bash e'



e.c
/*
 * a tiny expression evaluator
 * Copyright (C) 2014-2015  Yu-Jie Lin
 * Copyright (C) 2001  Dimitromanolakis Apostolos
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <float.h>
#ifndef BASH_LOADABLE
#include <getopt.h>
#endif
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

#ifdef BASH_LOADABLE
#include <builtins.h>
#include <shell.h>
#include <variables.h>
#include <bashgetopt.h>
#endif

#include "version.h"

bool DEBUG = false;
#define DPRINT(...)                     \
        if (DEBUG) {                    \
          fprintf(stderr, "DEBUG: ");   \
          fprintf(stderr, __VA_ARGS__); \
          fprintf(stderr, "\n");        \
        }
#define DPRINT_(...)   if (DEBUG) fprintf(stderr, __VA_ARGS__);

#define NAME    "e"
#define USAGE   NAME " [OPTIONS] [--] [expression]"

char *e_doc[] = {
  "Tiny expression evaluator.",
  "",
  "Options:",
  "",
  "  -h        display this help message",
  "  -V        display version string",
  "  -d        display debugging messages",
#ifdef BASH_LOADABLE
  "  -v VAR    store the result to VAR variable",
#endif
  (char *) NULL
};

#define COPYRIGHT \
        "Copyright (C) 2014-2015 Authors of e.bash\n"\
        "License GPLv2+: GNU GPL version 2 or later <https://www.gnu.org/licenses/gpl-2.0.html>\n"\
        "This is free software: you are free to change and redistribute it.\n"\
        "There is NO WARRANTY, to the extent permitted by law."

#define URL_HOMEPAGE  "https://bitbucket.org/livibetter/e.bash"
#define URL_ISSUES    URL_HOMEPAGE "/issues"

#ifndef BASH_LOADABLE
#define GETOPT()    getopt(argc, argv, "hVd")
#define OPTARG      optarg
#define FMTPNT(...) printf(__VA_ARGS__)
#define FORMAT(s)   format(s)
#define SUCCESS     EXIT_SUCCESS
#define FAILURE     EXIT_FAILURE
#define EXIT_USAGE  EXIT_FAILURE
#else
#define GETOPT()    internal_getopt(list, "hVdv:")
#define OPTARG      list_optarg
#define FMTPNT(...) sprintf(result + strlen(result), __VA_ARGS__)
#define FORMAT(s)   format(s, varname)
#define SUCCESS     EXECUTION_SUCCESS
#define FAILURE     EXECUTION_FAILURE
#define EXIT_USAGE  EX_USAGE
#endif


// copied from math.h
# define M_PI           3.14159265358979323846  /* pi */
# define M_E            2.7182818284590452354   /* e */


typedef double type;

// concatenated expression from argv or WORD_LIST
char *e;
char  c;
char *p;


type E();
/* For unknown reason, in Bash loadable with -O > 0:
 *
 * Running: e -- -1
 * Returns: -nan
 *
 * Even disabling all these optizmizers:
 *
 *   diff -y <(gcc -c -Q -O0 --help=optimizers) <(gcc -c -Q -O1 --help=optimizers) | grep '|'
 *
 * It doesn't help, only the following would get the correct answer.
 */
#ifdef BASH_LOADABLE
#pragma GCC push_options
#pragma GCC optimize ("O0")
#endif
type term();
#ifdef BASH_LOADABLE
#pragma GCC pop_options
#endif


void
next ()
{
  while ((c = *p++) == ' ');
}


void
syntax ()
{
  int tc = p - e - 1;
  int c;

  puts(e);

  if (tc >= 16) {
    for (c = 0; c < tc - 16; c++)
      putchar(' ');
    printf("syntax error ---^\n");
  } else {
    for (c = 0; c < tc; c++)
      putchar(' ');
    printf("^--- syntax error\n");
  }

  exit(FAILURE);
}


/**
 * mark the character for debugging messages
 */
void
debug_syntax (char *p)
{
  int tc = p - e;

  fprintf(stderr, "DEBUG: %s\n", e);
  fprintf(stderr, "DEBUG: %*s%s ", tc, "^", "---");
}
#define DEBUG_SYNTAX(p)  if (DEBUG) debug_syntax(p);

void
unknown (char *s)
{
  printf("'%s': unknown function\n", s);
  exit(FAILURE);
}


type
constant ()
{
  char *dp = p;
  type r = 0;

  while (c >= '0' && c <= '9') {
    r = 10 * r + (c - '0');
    next();
  }

  if (c == '.') {
    type p = 0.1;
    next();

    while (c >= '0' && c <= '9') {
      r += p * (c - '0');
      p /= 10;
      next();
    }
  }

  if (c == 'e' || c == 'E') {
    type m = 1;

    next();
    if (c == '-') {
      m = -m;
      next();
    } else {
      if (c == '+')
        next();
    }

    r *= pow(10, m * term());
  }

  DEBUG_SYNTAX(dp);
  DPRINT_("%f\n", r);
  return r;
}


type
function ()
{
  char f[20];
  char *dp = p;
  char *q;
  type v;

  q = f;
  while (q - f < 19 && c >= 'a' && c <= 'z') {
    *q++ = c;
    next();
  }

  *q = 0;

  #define RETURN(r)                       \
          do {                            \
            DEBUG_SYNTAX(dp);             \
            DPRINT_("%s = %f\n", f, r);   \
            return r;                     \
          } while (false);                \

  if (!strcmp(f, "pi"))
    RETURN(M_PI);
  if (!strcmp(f, "e" ))
    RETURN(M_E);

  if (!strcmp(f, "dblmax"))
    RETURN((type) DBL_MAX);
  if (!strcmp(f, "dblmin"))
    RETURN((type) DBL_MIN);
  if (!strcmp(f, "randmax"))
    RETURN((type) RAND_MAX);

  v = term();

  #define mathfunc(a,b)                         \
          if (!strcmp(f, a)) {                  \
            DEBUG_SYNTAX(dp);                   \
            DPRINT_("%s(%f) = %f\n", f, v, b);  \
            return b;                           \
          }
  #define mathfret(a,b,r)                             \
          if (!strcmp(f, a)) {                        \
            DEBUG_SYNTAX(dp);                         \
            DPRINT_("void %s(%f) = %f\n", f, v, r);   \
            b;                                        \
            return r;                                 \
          }

  mathfunc("sign"  , (v == 0) ? 0.0 : (v > 0) ? 1.0 : -1.0);

  mathfunc("abs"   , fabs(v));
  mathfunc("fabs"  , fabs(v));
  mathfunc("floor" , floor(v));
  mathfunc("ceil"  , ceil(v));
  mathfunc("round" , round(v));
  mathfunc("trunc" , trunc(v));
  mathfunc("sqrt"  , sqrt(v));
  mathfunc("exp"   , exp(v));

  mathfunc("sin"   , sin(v));
  mathfunc("cos"   , cos(v));
  mathfunc("tan"   , tan(v));
  mathfunc("asin"  , asin(v));
  mathfunc("acos"  , acos(v));
  mathfunc("atan"  , atan(v));

  mathfunc("sinh"  , sinh(v));
  mathfunc("cosh"  , cosh(v));
  mathfunc("tanh"  , tanh(v));
  mathfunc("asinh" , asinh(v));
  mathfunc("acosh" , acosh(v));
  mathfunc("atanh" , atanh(v));

  mathfunc("ln"    , log(v));
  mathfunc("log"   , log(v) / log(2));

  mathfunc("time"  , (type) time((time_t *) &v));
  mathfret("srand" , srand((unsigned int) v), (type) 0);
  mathfunc("rand"  , (type) rand());
  mathfunc("randf" , (type) rand() / (type) RAND_MAX);

  unknown(f);
  return 0;
}


type
term ()
{
  if (c == '(' || c == '[') {
    type r;

    next();
    r = E();
    if (c != ')' && c !=']')
      syntax();

    next();
    return r;
  } else if ((c >= '0' && c <= '9') || c == '.')
    return constant();
  else if (c >= 'a' && c <= 'z')
    return function();

  return 0;
}


static inline type
factorial (type v)
{
  type i;
  type r = 1;

  for (i = 2; i <= v; i++)
    r *= i;

  return r;
}


type
H ()
{
  char *dp = p;
  type q = term();
  type r = q;

  if (c == '!') {
    next();
    r = factorial(q);
    DEBUG_SYNTAX(dp);
    DPRINT_("%f! = %f\n", q, r);
  }

  return r;
}


type
G ()
{
  char *dp = p;
  type q;
  type r = H();

  while (c == '^') {
    next();
    q = G();
    r = pow(r, q);
    DEBUG_SYNTAX(dp);
    DPRINT_("^ %f = %f\n", q, r);
  }

  return r;
}


type
F ()
{
  char *dp;
  char  dc;
  type q;
  type r = G();

  while (true) {
    dp = p;
    dc = c;
    if (c == '*') {
      next();
      q = r;
      r *= G();
    } else if (c == '/') {
      next();
      q = r;
      r /= G();
    } else if (c == '%') {
      next();
      q = r;
      r = fmod(r, G());
    } else
      break;
    DEBUG_SYNTAX(dp);
    DPRINT_("%c %f = %f\n", dc, q, r);
  }

  return r;
}


type
E ()
{
  char *dp;
  char  dc;
  type q;
  type r = F();

  while (true) {
    dp = p;
    dc = c;
    if (c == '+') {
      next();
      q = r;
      r += F();
    } else if (c == '-') {
      next();
      q = r;
      r -= F();
    } else
      break;
    DEBUG_SYNTAX(dp);
    DPRINT_("%c %f = %f\n", dc, q, r);
  }

  return r;
}


type
S ()
{
  type r = E();

  if (c != 0)
    syntax();

  return r;
}


void
#ifndef BASH_LOADABLE
format (type X)
#else
format (type X, char *varname)
#endif
{
  type i;
  type f;
  int d;
#ifdef BASH_LOADABLE
  char result[2000] = "";
#endif

  if (!isfinite(X)) {
    FMTPNT("%f", X);
#ifdef BASH_LOADABLE
    goto out;
#endif
  }

  f = fabs(modf(X, &i));
  d = floor(log10(fabs(X))) + 1;

  if (isfinite(f) && f != 0) {
    char *p;
    char s[2000];
    char t[2000];

    sprintf(t, "%%.%df", 15 - d);
    sprintf(s, t, f);

    // remove all zeros from s
    p = s;
    while (*p)
      p++;
    p--;
    while (p > s && *p == '0')
      *p-- = 0;

    // decimal part has been rounded
    if (s[0] == '1')
      FMTPNT("%.0f", i + (X >= 0 ? 1 : -1));
    else {
      FMTPNT("%.0f", i);
      if (s[2] != 0)
        FMTPNT("%s", s + 1);
    }
  } else
    FMTPNT("%.0f", i);

#ifndef BASH_LOADABLE
  printf("\n");
#else
out:
  DPRINT("result is %s", result);
  if (varname) {
    bind_variable(varname, result, 0);
    DPRINT("Bound result to %s", varname);
  } else {
    printf("%s\n", result);
  }
#endif
}


/*
 * automatic memmory allocation for strcpy or re-allocation for strcat
 */
char *
strauto (char *dest, char *src) {
  if (dest == NULL) {
    if ((dest = malloc(strlen(src) + 1)))
      return strcpy(dest, src);

    fprintf(stderr, "error: malloc failed\n");
    exit(FAILURE);
  }

  if ((dest = realloc(dest, strlen(dest) + strlen(src) + 1)))
    return strcat(dest, src);

  fprintf(stderr, "error: realloc failed\n");
  exit(FAILURE);
}


void
print_help () {
  unsigned int i;

  puts("Usage: " USAGE);
  for (i = 0; i < sizeof(e_doc) / sizeof(char *); i++)
    if (e_doc[i])
      puts(e_doc[i]);

  printf("\n"\
         "Report bugs to <" URL_ISSUES ">\n"\
         "Home page: <" URL_HOMEPAGE ">\n");
}


int
#ifndef BASH_LOADABLE
main (int argc, char **argv)
#else
e_builtin (WORD_LIST *list)
#endif
{
  int opt;
#ifndef BASH_LOADABLE
  int i;
#else
  char *varname = NULL;

  /**
   * reset variables
   */
  DEBUG = false;
  e = NULL;
  p = NULL;
  c = 0;
#endif

#ifdef BASH_LOADABLE
  reset_internal_getopt();
#endif
  while ((opt = GETOPT()) != -1) {
    DPRINT("option = %c", opt);
    switch (opt) {
    case 'h':
      print_help();
      goto out;
    case 'V':
      puts(NAME " " VERSION);
      puts(COPYRIGHT);
      goto out;
    case 'd':
      DEBUG = true;
      DPRINT("debug is on");
      break;
#ifdef BASH_LOADABLE
    case 'v':
      varname = strauto(varname, OPTARG);
      DPRINT("varname = %s", varname);
      break;
#endif
    default:
#ifndef BASH_LOADABLE
      fprintf(stderr, "Usage: " USAGE "\n");
#endif
      return EXIT_USAGE;
    }
  }

  e = strauto(e, "");

#ifndef BASH_LOADABLE
  for (i = optind; i < argc; i++) {
    e = strauto(e, argv[i]);
    e = strauto(e, " ");
  }
#else
  list = loptend;
  while (list) {
    e = strauto(e, list->word->word);
    list = list->next;
    e = strauto(e, " ");
  }
#endif
  e[strlen(e) - 1] = '\0';
  p = e;

  srand(time(NULL));

  next();
  FORMAT(S());

  free(e);

out:
#ifdef BASH_LOADABLE
  free(varname);
#endif

  return SUCCESS;
}


#ifdef BASH_LOADABLE
struct builtin e_struct = {
  "e",
  e_builtin,
  BUILTIN_ENABLED,
  e_doc,
  USAGE,
  0
};
#endif
